Framework/HistogramData/test/RebinTest.h

```diff
@@ -0,0 +1,446 @@
+  }
+
+  void testRebinFailsStartBinEdgesInvalid() {
+    std::vector<double> binEdges{1, 1, 3, 4, 5, 7};
+    BinEdges edges(binEdges);
+
+    TS_ASSERT_THROWS(rebin(getCountsHistogram(), edges), std::runtime_error);
+  }
+
+  void testRebinEndCentralBinEdgesInvalid() {
+    std::vector<double> binEdges{1, 2, 3, 4, 5, 5};
+    BinEdges edges(binEdges);
+
+    TS_ASSERT_THROWS(rebin(getCountsHistogram(), edges), std::runtime_error);
+  }
+
+  void testRebinFailsInputBinEdgesInvalid() {
+    std::vector<double> binEdges{1, 2, 3, 3, 5, 7};
+    Histogram hist(BinEdges(std::move(binEdges)), Counts(5, 10));
+    BinEdges edges{1, 2, 3, 4, 5, 6};
+
+    TS_ASSERT_THROWS(rebin(hist, edges), std::runtime_error);
+  }
+
+  void testRebinFailsInputAndOutputBinEdgesInvalid() {
+    std::vector<double> binEdges{1, 2, 3, 3, 5, 7};
+    Histogram hist(BinEdges(binEdges), Counts(5, 10));
+    BinEdges edges(std::move(binEdges));
+
+    TS_ASSERT_THROWS(rebin(hist, edges), std::runtime_error);
+  }
+
+  void testRebinIdenticalBins() {
+    auto histCounts = getCountsHistogram();
+    auto histFreq = getFrequencyHistogram();
+
+    auto outCounts = rebin(histCounts, histCounts.binEdges());
+    auto outFreq = rebin(histFreq, histFreq.binEdges());
+
+    TS_ASSERT_EQUALS(outCounts.x().rawData(), histCounts.x().rawData());
+    TS_ASSERT_EQUALS(outCounts.y().rawData(), histCounts.y().rawData());
+    TS_ASSERT_EQUALS(outCounts.e().rawData(), histCounts.e().rawData());
+
+    TS_ASSERT_EQUALS(outFreq.x().rawData(), histFreq.x().rawData());
+    TS_ASSERT_EQUALS(outFreq.y().rawData(), histFreq.y().rawData());
+    TS_ASSERT_EQUALS(outFreq.e().rawData(), histFreq.e().rawData());
+  }
+
+  void testBinEdgesOutsideInputBins() {
+    auto histCounts = getCountsHistogram();
+    auto histFreq = getFrequencyHistogram();
+
+    auto outCounts = rebin(histCounts, BinEdges(10, LinearGenerator(30, 1)));
+    auto outFreq = rebin(histFreq, BinEdges(10, LinearGenerator(30, 2)));
+
+    TS_ASSERT(std::all_of(outCounts.y().cbegin(), outCounts.y().cend(),
+                          [](const double i) { return i == 0; }));
+    TS_ASSERT(std::all_of(outFreq.y().cbegin(), outFreq.y().cend(),
+                          [](const double i) { return i == 0; }));
+    TS_ASSERT(std::all_of(outFreq.e().cbegin(), outFreq.e().cend(),
+                          [](const double i) { return i == 0; }));
+  }
+
+  void testSplitBinSymmetric() {
+    // Handles the case where
+    // | | |   becomes:
+    // |||||
+    Histogram hist(BinEdges{0, 1, 2}, Counts{10, 10});
+    Histogram histFreq(BinEdges{0, 1, 2}, Frequencies{12, 12});
+    BinEdges edges{0, 0.5, 1, 1.5, 2};
+
+    auto outCounts = rebin(hist, edges);
+    auto outFreq = rebin(histFreq, edges);
+
+    for (size_t i = 0; i < outCounts.y().size(); i++) {
+      TS_ASSERT_EQUALS(outCounts.y()[i], 5.0);
+      TS_ASSERT_DELTA(outCounts.e()[i], std::sqrt(5), 1e-5);
```

Comment:
It makes a big difference. Someone might break `rebin` without breaking tests if the tolerances are so large. So always make bounds as tight as feasible.